File: hw1/hw1.Tests/Parser.Tests.cs
using hw1;
using Xunit;

namespace Tests
{
    public class ParserTests
    {
        [Fact]
        public void WhenArgsLengthIncorrect()
        {
            Check(new[]{"1", "+"}, false, Parser.WrongArgsLength);
            Check(new[]{"1", "+", "2", "="}, false, Parser.WrongArgsLength);
        }

        [Fact]
        public void WhenOperationIncorrect()
        {
            Check(new[]{"1", "%", "2"}, false, Parser.WrongOperation);
        }

        [Fact]
        public void WhenArgsIncorrect()
        {
            Check(new[]{"x", "+", "1"}, false, Parser.WrongArgFormat);
            Check(new[]{"1", "+", "x"}, false, Parser.WrongArgFormat);
        }

        [Fact]
        public void WhenAllCorrect()
        {
            Check(new[]{"1", "+", "2"}, true, Parser.AllCorrect, 1, Operation.Plus, 2);
            Check(new[]{"1", "-", "2"}, true, Parser.AllCorrect, 1, Operation.Minus, 2);
            Check(new[]{"1", "*", "2"}, true, Parser.AllCorrect, 1, Operation.Multiply, 2);
            Check(new[]{"1", "/", "2"}, true, Parser.AllCorrect, 1, Operation.Divide, 2);
        }

        private void Check(string[] args, bool expectedResult, int expectedCode, int expectedVal1, Operation expectedOperation, int expectedVal2)
        {
            var result = Parser.TryParseArgs(args, out var val1, out var operation, out var val2, out var code);
            Assert.Equal(expectedResult, result);
            Assert.Equal(expectedCode, code);
            Assert.Equal(expectedVal1, val1);
            Assert.Equal(expectedOperation, operation);
            Assert.Equal(expectedVal2, val2);
        }

        private void Check(string[] args, bool expectedResult, int expectedCode)
        {
            var result = Parser.TryParseArgs(args, out _, out _, out _, out var code);
            Assert.Equal(expectedResult, result);
            Assert.Equal(expectedCode, code);
        }
    }
}

File: hw4/hw4.Tests/Parser.Tests.fs
module Tests.ParserTests

open System
open Xunit
open hw4

[<Fact>]
let ``ParseOperation simple test`` () =
    Assert.Equal(Operation.Plus, Parser.parseOperation "+")
    Assert.Equal(Operation.Minus, Parser.parseOperation "-")
    Assert.Equal(Operation.Multiply, Parser.parseOperation "*")
    Assert.Equal(Operation.Divide, Parser.parseOperation "/")

[<Fact>]
let ``ParseOperation should throws WrongOperation`` () =
    let parseWrongOperation () =
        let a = Parser.parseArgs [|"1";"%";"2"|]
        printf $"%s{a.ToString()}"
    let action = Action parseWrongOperation
    Assert.Throws<Exceptions.WrongOperation>(action)

[<Fact>]
let ``ParseArgs simple test`` () =
    Assert.Equal((1., Operation.Plus, 2.), Parser.parseArgs [| "1";"+";"2" |])

[<Fact>]
let ``ParseArgs should throws WrongArgsLength`` () =
    let parseTwoArgs () =
        let a = Parser.parseArgs [|"1";"+"|]
        printf $"%s{a.ToString()}"
    let action = Action parseTwoArgs
    Assert.Throws<Exceptions.WrongArgsLength>(action)

[<Fact>]
let ``ParseArgs should throws WrongArgFormat`` () =
    let parseWrongArg1 () =
        let a = Parser.parseArgs [|"x";"+";"1"|]
        printf $"%s{a.ToString()}"
    let action1 = Action parseWrongArg1
    let _ = Assert.Throws<Exceptions.WrongArgFormat>(action1)

    let parseWrongArg2 () =
        let a = Parser.parseArgs [|"1";"+";"x"|]
        printf $"%s{a.ToString()}"
    let action2 = Action parseWrongArg2
    Assert.Throws<Exceptions.WrongArgFormat>(action2)